yente/search/base.py

```diff
@@ -74,3 +91,317 @@ async def close_es() -> None:
+            case "ADD":
+                return self._create_operation(entity, doc_id, index)
+            case "MOD":
+                return self._update_operation(entity, doc_id, index)
+            case "DEL":
+                return self._delete_operation(doc_id, index)
+            case _:
+                raise ValueError(f"Unknown operation type: {body.get('op')}")
+
+
+class ESSearchProvider(SearchProvider):
+    @classmethod
+    async def create(cls) -> "ESSearchProvider":
+        self = cls()
+        self.client = await get_es()
+        return self
+
+    async def upsert_index(self, index: str) -> None:
+        """
+        Create an index if it does not exist. If it does exist, do nothing.
+        """
+        try:
+            schemata = list(model.schemata.values())
+            mapping = make_entity_mapping(schemata)
+            await self.client.indices.create(
+                index=index, mappings=mapping, settings=INDEX_SETTINGS
+            )
+        except BadRequestError:
+            pass
+
+    async def _remove_write_block(self, index: str) -> None:
+        await self.client.indices.put_settings(
+            index=index, settings={"index.blocks.read_only": False}
+        )
+
+    async def _add_write_block(self, index: str) -> None:
+        await self.client.indices.put_settings(
+            index=index, settings={"index.blocks.read_only": True}
+        )
+
+    async def clone_index(self, index: str, new_index: str) -> None:
+        try:
+            await self._add_write_block(index)
+            await self.client.indices.clone(
+                index=index,
+                target=new_index,
+                body={
+                    "settings": {"index": {"blocks": {"read_only": False}}},
+                },
+            )
+        finally:
+            await self._remove_write_block(index)
+
+    async def delete_index(self, index: str) -> None:
+        """
+        Delete a given index if it exists.
+        """
+        try:
+            await self.client.indices.delete(index=index)
+        except NotFoundError:
+            pass
+
+    async def get_backing_indexes(self, name: str) -> List[str]:
+        resp = await self.client.indices.get_alias(name=name)
+        return list(resp.keys())
+
+    async def index_exists(self, index: str) -> bool:
+        exists = await self.client.indices.exists(index=index)
+        if exists.body:
+            log.info("Index is up to date.", index=index)
+            return True
+        return False
+
+    async def rollover(self, alias: str, new_index: str, prefix: str = "") -> None:
+        """
+        Remove all existing indices with a given prefix from the alias and add the new one.
+        """
+        actions = []
+        actions.append({"remove": {"index": f"{prefix}*", "alias": alias}})
+        actions.append({"add": {"index": new_index, "alias": alias}})
+        await self.client.indices.update_aliases(actions=actions)
```

Comment:
Making this fully atomic is hella cool

Reply:
Yeah, it's a cool way of doing index migration tbh